[JBPM-7974] kie-server not removed from the controller view if server  is killed or crashes

elguardian opened the pull request:
fixes some problem during refreshing in the UI
added alive ping to check whether servers are alive

elguardian commented:
ok to test

krisv commented:
@porcelli this is an update to the kie-server UI, can someone review this?

porcelli commented:
@krisv @elguardian the rule of thumb is to always provide tests, so please provide some to cover the changes.

jomarko commented:
Hi @krisv, @elguardian I can do QE review for this, however as have been said, can not approve until unit tests are added.

elguardian commented:
hi @jomarko 
I have implemented the test case. I didn't see any example in the kiegroup implemented an IT scenario where one of the nodes gets kills  so I implemented the ServerKillProcessor in arquillian (only works in linux)

The scenario is 
2 nodes (1 deploy the workbench and 1 deploying the kie-server)
check the number of instances = 1
invoking kill container in arquillian 
(wait for the polling mechanism to kick in through the count down latch)
check the number of instance = 0

let me know if you have questions.

jomarko commented on kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-backend/src/main/java/org/kie/workbench/common/screens/server/management/backend/KieServerControllerStartup.java:
```diff
@@ -0,0 +1,50 @@
+    private RestKieServerHealthCheck checker;
+
+    @Inject
+    private RestKieServerControllerImpl kieServer; // just the dependency
```
Could you please explain more, why we need to inject `RestKieServerControllerImpl`? I mean I don't see it is used anyhow.

jomarko commented on kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-backend/src/main/java/org/kie/workbench/common/screens/server/management/backend/rest/RestKieServerHealthCheck.java:
```diff
@@ -0,0 +1,138 @@
+        try {
+            PING_INTERVAL = Long.parseLong(System.getProperty("kie.ping.alive.timeout", "5000"));
+        } catch (NumberFormatException e) {
+            logger.warn("The property " + PING_ALIVE_PROPERTY + " is not a number; Fallback to 1000 ms ping");
```
You log fallback to `1000 ms` but set to `5000 ms` on line below, please fix either log or the value.

jomarko replied:
@elguardian thank you for changes, could you please have a look also on this comment, otherwise I think we are fine.

jomarko commented on kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-backend/src/main/java/org/kie/workbench/common/screens/server/management/backend/rest/RestKieServerHealthCheck.java:
```diff
@@ -0,0 +1,138 @@
+            PING_DISABLED = Boolean.parseBoolean(System.getProperty("kie.ping.alive.disable", "false"));
+        } catch (NumberFormatException e) {
+            logger.warn("The property " + PING_ALIVE_DISABLED + " is not true/false value; Fallback to false");
+            PING_INTERVAL = 5000L;
```
I think this fallback should set `PING_DISABLED` instead of `PING_INTERVAL`, could you please double check?

jomarko commented on kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-backend/src/test/java/org/kie/workbench/common/screens/server/management/backend/arquillian/DefaultLinuxKillProcessor.java:
```diff
@@ -0,0 +1,55 @@
+            killSequence = "kill -9 `ps aux | grep -v 'grep' | grep 'jboss.home.dir=[jbossHome] ' | sed -re '1,$s/[ \\t]+/ /g' | cut -d ' ' -f 2`";
+            killSequence = killSequence.replace("[jbossHome]", container.getContainerConfiguration().getContainerProperties()
+                                                                        .get("jbossHome"));
+            p = Runtime.getRuntime().exec(new String[]{"sh", "-c", killSequence});
```
Could you please double check if the `Process` instance `p` is released properly? I am not saying there is an issue, but locally I am using the `infer` tool and got this message:
```
kie-wb-common-server-ui-backend/src/test/java/org/kie/workbench/common/screens/server/management/backend/arquillian/DefaultLinuxKillProcessor.java:40: error: RESOURCE_LEAK
  resource of type `java.lang.Process` acquired to `p` by call to `exec(...)` at line 39 is not released after line 40.
**Note**: potential exception at line 40
```

elguardian commented:
@jomarko @porcelli  any progress with the review

porcelli commented:
@jomarko? @cristianonicolai could you also take a look?

cristianonicolai commented:
@elguardian overall looks good but I was wondering about the standalone controller as well, I believe that scenario is not covered in here correct?
Also, REST communication with the Kie Server is not always possible as it could be behind a load balancer, that's why we added websocket support, have you considered using that to check if the server still alive?

jomarko commented on kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-backend/src/test/java/org/kie/workbench/common/screens/server/management/backend/AbstractControllerIT.java:
```diff
@@ -122,6 +115,9 @@ public static WebArchive createWorkbenchWar() {
                 .addClass(AbstractControllerIT.class)
+                                             .addClass(AbstractAutoControllerIT.class)
+                                             .addClass(RestKieServerHealthCheck.class)
+                                             .addClass(KieServerControllerStartup.class)
```
Please align these thrre lines with same indent as other `addClass` calls.

jomarko commented:
@elguardian Sorry, I read a [comment](https://github.com/kiegroup/kie-wb-common/pull/2285#issuecomment-457404286) after I submitted review, if it means code will change, please let me do re-review then.

elguardian commented:
@cristianonicolai 
WS uses a permantent connection so it should detect an OS level event like killing a socket (jvm death). 
A different topic would be if an intermediate hw like a LB or FW kills the connection because inactivity and severe the connection while the socket remains open in both ends of communication (kie-server/kie-controller). What will happen in this case is that if you try to write you will get the infamous "connection reset" from the FW/LW forcing the socket to close. (this problem is not treated here)
I haven't thought about standalone controller.

elguardian commented:
@cristianonicolai did the previous comment answer your question ?

cristianonicolai commented:
@elguardian I'm fine if you prefer to check the standalone controller on a different PR. Regarding WS, I was just wondering if in this same scenario (os kill) the server instance is also properly removed from the controller.

elguardian commented:
@cristianonicolai updated and added for standalone controller

this depends on
https://github.com/kiegroup/droolsjbpm-integration/pull/1712

elguardian commented:
rebased

elguardian commented:
ok to test

elguardian commented:
a bunch or marshalling errors... not really sure where they come from.

elguardian commented:
jenkins please retest this

elguardian commented:
failures not related

elguardian commented:
rebased
@cristianonicolai  @mswiderski is there anything missing here ? i would like to backport this for 7.3.1 if possible.

mswiderski commented:
Jenkins retest this

cristianonicolai commented:
@elguardian you probably need to rebase your PR now. Kie Server mode REGULAR has been removed and it is causing the build to fail.

elguardian commented:
@cristianonicolai changed in integration project. should compile now

elguardian commented:
Jenkins retest this

Rikkola commented:
The master build can't find the KieServerHealthCheckControllerImpl.java. Is there another PR or commit I need? 

Could be I'm out of sync with the builds.

elguardian commented:
@Rikkola yeah there is
https://github.com/kiegroup/droolsjbpm-integration/pull/1712

cristianonicolai commented:
@elguardian I missed that dependency in the comments, perhaps we should make sure that any dependency is added into the initial comment/description.

elguardian commented:
@cristianonicolai duly noted
